refactor: foss event and project grants' styles and workflows

RiyaMathew-11 opened the pull request:
## What type of PR is this? (check all applicable)
- [x] 🍕Feature
- [x] ⚙️Chore
- [x] 🐛Bug Fix
- [x] 🧑‍💻Refactor

## Description
Fixed and refactored the event and project grants workflows, templates, style cleanups and a small feat. 

-  Fixed bugs in event grant data display on webpage while migrating old records to new site.
   - Added fields for custom grant amount 
   - Logic to handle missing values.
   - Adjusted settings for list view and filter for project and event grant doctypes

-  Added functions that loads and displays more and less event grant data on webpage when required, on a button click.
   - The page displays 8 entries by default when loaded. 
   - The JS functions `loadMoreGrants` and `viewLessGrants` acts to load and display event grant data in the form of cards. 
   - `get_more_grants` methods works as an api that fetches grant data from the doctype.

- Removed extra unwanted css from the web templates making up **Funding and Grants Page** and changed all hard-coded text-style values to variables
   - Added color property to hover for primary and secondary button in [custom.css](https://arc.net/l/quote/jshixcic) to fix broken styles [when the class is used with `a` tags](https://arc.net/l/quote/umujmjif). 

- Cleaned up all recurring .py files that had server-side api methods, rewrote them to a single [pages.py](https://github.com/fossunited/fossunited/commit/438928a06f5f6ccc1871b925610b2b1b956124f5#diff-2f1b237a4e705c0ceeb45993087a5a7e32ad29017adb80fcef6980af243d0ae0) file and updated the routes where these methods are used. 
  - Includes the `search_foss_club()` from the FOSS Clubs template



## Related Issues & Docs

- Closes #300 
- Closes #303 

## Checklist
- [x] I have read the [contribution guidelines]("./docs/CONTRIBUTING.md").
- [x] I have tested these changes locally.
- [x] I have updated the documentation (If applicable).
- [x] The code follows the project's coding standards.
- [x] I have added/updated tests, if applicable.

## Screenshots/GIFs/Screen Recordings 

![Screenshot 2024-05-14 at 12 18 10 AM](https://github.com/fossunited/fossunited/assets/60656677/76a1b1a2-d853-4c67-988b-e3d058074bda)


The event and project grant sections: 

![grants-demo](https://github.com/fossunited/fossunited/assets/60656677/469c25b0-7d8f-4f4a-9f96-217f13590d51)

rahulporuri commented:
Cross-posting the PR writing/review-related references that I shared internally
* https://google.github.io/eng-practices/
* https://mtlynch.io/code-review-love/
* https://mtlynch.io/human-code-reviews-1/
* https://mtlynch.io/human-code-reviews-2/

I have a meta-comment before I start my review - is it possible to "deploy" the changes from this PR to a test site? The test site makes life a bit easier for reviewers. The alternative is to setup frappe bench, checkout this branch and deploy it locally - not a big deal but the test site save a little bit of time. I've seen similar solutions for docs changes in FOSS projects where a temporary PR-specific ReadTheDocs page gets created.

Tangential but related - I added the two issues that will be closed by this PR (#300 , #303 ) to the GitHub project and moved them to "In Progress". You had already assigned yourself to the issues, which is helpful!

1. The scope of the PR is confined, making it easy to review
2. The PR is also small in terms of the changelist, again making it easy to review
3. The description doesn't need to highlight the specific commits and the changelist in the commits, just the information is sufficient. **BUT** it is beneficial to the PR review process when the commits are self-contained and there is less to no cross-contamination (for lack of a better word) between the commits. Each commit can then be reviewed separately, reducing the amount of context a reviewer needs to hold in their head
4. The screenshots are helpful but a screen capture would have been better because the changes in interactivity aren't obvious from the screenshots. I've seen the team use a few different screen capture software, most of them online. I use [Licecap by Cockos](https://www.cockos.com/licecap/). It's a small installable and it works offline. It creates GIFs that can be uploaded to GitHub.

That's the big picture review. I'll dig into the code and potentially make nitpicky comments separately.

rahulporuri commented on fossunited/api/pages.py:
```diff
@@ -0,0 +1,43 @@
+
+
+@frappe.whitelist(allow_guest=True)
+def search_foss_club(query):
```
Q. I'm seeing type hints in a few places in the code base but not in others. It looks like Frappe provides type hints. Have we defined the code standards for the project?

At my previous employer, we wanted to experiment with type hints, so we added them incrementally and evaluated regularly if they continue to provide value. We never retroactively updated the entire code base to include type hints because that is a huge investment.

RiyaMathew-11 replied:
We haven't promptly decided/mandated yet to use type hints - and I haven't used it yet in any functions that I have. I think it seems seems to be used [impromptu in places](https://github.com/fossunited/fossunited/blob/develop/fossunited/api/dashboard.py#L10) where it has been used, based on preference.

@NagariaHussain what are your thoughts? Should we start the practice? 

> Have we defined the code standards for the project?

We don't have a specific document perse nor mandated, but have been doing improvements with feedback and suggestions. I feel it would be good to have one too.

rahulporuri commented on fossunited/api/pages.py:
```diff
@@ -0,0 +1,43 @@
+
+
+@frappe.whitelist(allow_guest=True)
+def search_foss_club(query):
```
Whether or not we decide to use type hints, in my humble opinion docstrings are a must.

What the function does is pretty evident from the function name (which is great) but it's not clear to me what "query" parameter is and examples of what it could be. The same goes for `club_list`, it looks like a list but it's not clear what the list element types are.

Update: This PR introduces other functions and we should consider adding docstrings to all of them.

rahulporuri replied:
Relevant resources
* https://peps.python.org/pep-0257/ the PEP that describes docstring conventions in Python
* https://www.python.org/dev/peps/pep-0008/ the Python style guide
* https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard NumPy style guide for docstrings. Especially relevant is the section https://numpydoc.readthedocs.io/en/latest/format.html#parameters on documenting parameters. We heavily relied on it at my previous employer.

I'm not sure what the Frappe docstring conventions are

RiyaMathew-11 replied:
I agree with the use of docstrings too - I feel we should go for writing docstrings to all the existing functions now in a cleanup commit - maybe starting with the api methods ?

Thoughts @harshtandiya @NagariaHussain  ?

rahulporuri commented on fossunited/api/pages.py:
```diff
@@ -0,0 +1,43 @@
+
+
+@frappe.whitelist(allow_guest=True)
+def search_foss_club(query):
```
it looks like the function returns a list of 0 or more clubs so how does `search_foss_clubs` (plural) sound as the function name?

Nitpicky comment - On second thought, how does `get_foss_clubs` sound?
Note: Nitpicky comments don't have to be addressed. They usually convey personal opinion and/or style.

RiyaMathew-11 replied:
The function actually searches the doctype based on the given query in the search-component and returns the foss clubs as list based on it. 

Maybe we can make it `search_foss_clubs()` ? as it returns 0 or a list of clubs upon search. Since this change is out of scope in terms of this PR - i think it would be better to change it in a different PR too.

rahulporuri replied:
`search_foss_clubs` sounds good to me. resolving this conversation.

rahulporuri commented on fossunited/api/pages.py:
```diff
@@ -0,0 +1,43 @@
+            "chapter_type": "FOSS Club",
+        },
+        or_filters=[
+            ["city", "like", "%" + query + "%"],
+            ["chapter_name", "like", "%" + query + "%"],
```
nitpicky comment - I like `f'strings'`

```suggestion
            ["city", "like", f"%{query}%"],
            ["chapter_name", "like", f"%{query}%"],
```

RiyaMathew-11 replied:
I agree too with it's usage. I will fix them in my next PR of bug fixes related to FOSS Clubs and IP. This addition I think would be a bit out of scope with this PR.

rahulporuri commented on fossunited/chapters/web_template/foss_clubs_home_page/foss_clubs_home_page.html:
Tangential comment - What formatters do we use? It looks like we use Ruff to format Python code but we also have JSON and HTML code in the codebase. Does Frappe recommend any formatters for the HTML and JSON code?

RiyaMathew-11 replied:
Hussain would know better about this. @NagariaHussain can you share your thoughts ?

rahulporuri reviewed (commented):
Mostly LGTM (Looks Good To Me). I've added a number of comments, including nitpicky comments on the Python code. I don't know enough HTML and my CSS knowledge is next to none so I didn't comment on that.

Most of my comments aren't Frappe-specific and that will hopefully change over time.

rahulporuri commented:
> - [x] I have added/updated tests, if applicable.

It doesn't look like there are new tests in this PR nor are any existing tests updated. _I don't know if we have tests at all in the code base_ Again, I don't know enough Frappe to provide meaningful comments on how easy or hard it will be to write unittests for the changes introduced in this PR.

RiyaMathew-11 committed:
chore: add custom grant amount field and adjusted list view and filter settings

RiyaMathew-11 committed:
feat: view more - less grants and logic to handle missing values

RiyaMathew-11 committed:
refactor: cleanup styles and html in grants' templates

RiyaMathew-11 committed:
chore: rewrote all api methods to pages.py

RiyaMathew-11 commented:
> Cross-posting the PR writing/review-related references that I shared internally
> 
> * https://google.github.io/eng-practices/
> * https://mtlynch.io/code-review-love/
> * https://mtlynch.io/human-code-reviews-1/
> * https://mtlynch.io/human-code-reviews-2/
> 
> I have a meta-comment before I start my review - is it possible to "deploy" the changes from this PR to a test site? The test site makes life a bit easier for reviewers. The alternative is to setup frappe bench, checkout this branch and deploy it locally - not a big deal but the test site save a little bit of time. I've seen similar solutions for docs changes in FOSS projects where a temporary PR-specific ReadTheDocs page gets created.
> 
> Tangential but related - I added the two issues that will be closed by this PR (#300 , #303 ) to the GitHub project and moved them to "In Progress". You had already assigned yourself to the issues, which is helpful!
> 
> 1. The scope of the PR is confined, making it easy to review
> 2. The PR is also small in terms of the changelist, again making it easy to review
> 3. The description doesn't need to highlight the specific commits and the changelist in the commits, just the information is sufficient. **BUT** it is beneficial to the PR review process when the commits are self-contained and there is less to no cross-contamination (for lack of a better word) between the commits. Each commit can then be reviewed separately, reducing the amount of context a reviewer needs to hold in their head
> 4. The screenshots are helpful but a screen capture would have been better because the changes in interactivity aren't obvious from the screenshots. I've seen the team use a few different screen capture software, most of them online. I use [Licecap by Cockos](https://www.cockos.com/licecap/). It's a small installable and it works offline. It creates GIFs that can be uploaded to GitHub.
> 
> That's the big picture review. I'll dig into the code and potentially make nitpicky comments separately.

Thank you so much for the review @rahulporuri! Based on this, I've made some changes to the PR:

- Upon reading the material, I've updated the PR description, to make it more contextual and crisp as it can be made. 
- I've re-arranged the commits from 7 to 4 - to make them more easier to navigate as well.

RiyaMathew-11 commented:
> > * [x]  I have added/updated tests, if applicable.
> 
> It doesn't look like there are new tests in this PR nor are any existing tests updated. _I don't know if we have tests at all in the code base_ Again, I don't know enough Frappe to provide meaningful comments on how easy or hard it will be to write unittests for the changes introduced in this PR.

We dont have tests written on that note, and they aren't needed in this PR as well, it maybe be added to the PR only if applicable. 

In context of unit-tests with Frappe, I don't have enough knowledge on it as well. Perhaps @NagariaHussain can help by sharing his thoughts on this ?

rahulporuri reviewed (approved):
Changes LGTM but like I mentioned. If there are no other reviews by EoD tomorrow, please go ahead and merge.

Let's also open the relevant issues for what has been discussed in the PR review before considering this PR "complete".

harshtandiya commented on fossunited/fossunited/doctype/foss_event_grant/foss_event_grant.json:
```diff
@@ -149,8 +154,9 @@
   {
    "fieldname": "grant_amount",
    "fieldtype": "Select",
+   "in_standard_filter": 1,
    "label": "Grant amount",
-   "options": "\nRs. 10,000\nRs. 25,000\nRs. 50,000"
+   "options": "\nRs. 10,000\nRs. 25,000\nRs. 50,000\nCustom"
   },
```
Why don't we have just a custom amount field rather than having a select, then custom. We can just put the amount directly. It's going to be used internally anyway

RiyaMathew-11 replied:
Grant amount is usually mostly fixed for events: 

- 10,000 K for hackathons
- 25,000 K for mini-confs 
- 50,000 K for major confs 

It goes rarely out of this scope, and can the grant amount category is attributed to the type of event as well. The custom amount is to actually handle edge cases.

harshtandiya commented on fossunited/fossunited/web_template/recent_event_grants/recent_event_grants.html:
```diff
@@ -175,6 +161,97 @@ <h6>{{grant.grant_amount}}</h6>
+        document.getElementById('view-more').style.display = 'block';
+    }
+
+    function formatGrantHTML(grant) {
+        return `
+            <h6>${grant.event_name}</h6>
+
+            <div class="event-detail" id="event-date">
+                <p>Date</p>
+                <p class="grant-card-value">${grant.event_start_date ? new Date(grant.event_start_date).toLocaleDateString('en-US', { day: '2-digit', month: 'short', year: 'numeric' }) : "TBA"}</p>
+            </div>
+
+            <div class="event-detail" id="event-location">
+                <p>Location</p>
+                <p class="grant-card-value">${grant.event_location || "TBA"}</p>
+            </div>
+
+            <div class="grant-card-footer">
+                <div class="event-detail">
+                    <p>Grant Amount</p>
+                    <p class="grant-card-value" id="grant-amount">${grant.grant_amount !== "Custom" ? grant.grant_amount : grant.custom_amount}</p>
```
Where is this function used and what does this do?

RiyaMathew-11 replied:
It's used in the [`loadMoreGrants()`](https://arc.net/l/quote/llbswnmu) and is used to make the fetched grant data formatted as cards and display.
